smart_open/smart_open_lib.py

```diff
@@ -381,12 +381,36 @@ def _open_binary_stream(uri, mode, **kw):
         if host is not None:
             kw['endpoint_url'] = _add_scheme_to_host(host)
         return smart_open_s3.open(uri.bucket.name, uri.name, mode, **kw), uri.name
-    elif hasattr(uri, 'read'):
+    elif _is_stream(uri, mode):
         # simply pass-through if already a file-like
-        filename = '/tmp/unknown'
+        # we need to return something as the file name, but we don't know what
+        # so we probe for uri.name (e.g., this works with tempfile.NamedTemporaryFile)
+        # if the value ends with COMPRESSED_EXT, we will note it in _compression_wrapper()
+        # if there is no such an attribute, we return "unknown" - this effectively disables
+        # any compression (the actual string does not matter ofc)
+        filename = getattr(uri, 'name', 'unknown')
         return uri, filename
     else:
-        raise TypeError('don\'t know how to handle uri %s' % repr(uri))
+        raise TypeError("don't know how to handle uri %r in mode %r" % (uri, mode))
+
+
+def _is_stream(fileobj, mode):
```

Comment:
Good observation! I will change the function to check `readable` and `writeable` attribute values instead, if they are present.

There is also an issue with `seek()`, which may be required by some compression codecs. I have not had time to dig it deeply yet, but from my experience, this may be a real compatibility problem.

Reply:
Besides, regarding `read` and `write`, we can pass any custom class without some methods implemented. Pretty common practice with some streams derived from networking. After a second thought, I tend to leave my checks, but also improve the compatibility with the standard streams.

Reply:
Personally, I think it's better off to **not** check those things at all. This is both more Pythonic (duck typing, asking for forgiveness instead of permission, and a couple of other things) and provides more clarity to the user.

With this function, if a required method is missing or unsupported, the user will get a general TypeError:

    "don't know how to handle uri %r in mode %r"

Now, if we were a user application, that would be a decent error message. We're not an application, though, we're a library, so the above message is less helpful than the alternative.

The alternative is to let smart_open try to read/write from the stream, and if something is missing or unimplemented, allow it to raise the appropriate exception. That will tell the user of our library exactly **why** we can't handle the stream they gave us. A good example is the seek scenario - if a user gives us an unseekable stream and we absolutely **must** seek it, they'll get the precise reason why we can't process it, not a generic error message.

Reply:
Duck typing does not hinder user-friendly assertions before a shipwreck happens. In my experience, getting an AttributeError somewhere on a 30th place in the call stack does not really help users to figure out what went wrong. It is often the opposite - they have no idea what is going on in the implementation details and they are going to open an issue on GitHub to ask if there is a bug in the library (true story, sadly).

Thus if we are sure that some particular methods or modes will be used in the future - we should check them *before* bumming an unfriendly stack trace. At the same time, as long as we don't mess with `isinstance`, our duck typing quacks.

You've got a valid concern regarding the usefulness of the generic `TypeError`. Indeed, I am going to change the code to raise proper errors which point to the exact problems.

Reply:
While I think the PR is good in general, I'm **strongly** opposed to _this_ particular change for the following reasons:

1. Little practical value
2. Lack of correctness
3. Introduction of unnecessary coupling
4. Disproportionate maintenance overhead

I explain each of these points in more detail below.

---

First, the new code adds little practical value. The existing code (checking for the read attribute) already handles 99% of the use case correctly: as I've already mentioned, all streams from the standard library have a read method, regardless of whether they're actually readable. The existing code gives informative and correct information via exception handling. For example, here's an attempt to write to a read-only stream:

```
$ python -c 'import smart_open as s;s.smart_open(open("README.rst")).write("no")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
io.UnsupportedOperation: not writable
```

> In my experience, getting an AttributeError somewhere on a 30th place in the call stack does not really help users to figure out what went wrong.

This is not a valid argument. Both the new and old code raise exceptions at similar levels in the call stack, so the experience for the user will be similar.

Attempts to improve on the existing 99% are subject to diminishing returns. More specifically, the new code is spending an unnecessary amount of effort preparing for edge cases that do not occur in practice. This is why the new unit tests need custom buffer classes: no I/O classes in the Python standard library behave in the way the new code and tests expect streams to behave. It's like boiling the ocean to make a cup of tea.

---

Second, the new code still does not achieve its intended goal: successfully checking whether an input parameter is a stream that we can work with. You've checked for presence of the read/write function, but that isn't enough:

1. Is the stream actually [readable/writable](https://docs.python.org/3.5/library/io.html#io.IOBase.readable)?
1. Does the stream support seeking? Does it support seeking for particular values of the `whence` parameter (e.g. seek from end of file)?
2. Is the stream actually open?
3. Is it a binary or unicode stream? Is that compatible with the mode we're working with?
4. ... and much more

I'm not proposing you actually implement checks for the above. I'm telling you that the only way to 100% effectively and correctly determine whether you can peform an I/O operation on a stream (e.g. read or write) is to actually try it and see.

Also, the exception raised by the new code is misleading. TypeError suggests there is a problem with the _type_, which isn't really the case here.

---

Third, the new code introduces unnecessary coupling between input validation and the actual implementation.

> Thus if we are sure that some particular methods or modes will be used in the future - we should check them **before** bumming an unfriendly stack trace.

This is a bad idea. The new validation code essentially needs to know details of the implementation. You've introduced coupling: if the implementation changes in the future, we'll need to remember to update the validation code. What's worse, this coupling is _hidden inside_ the new input validation function. So now, whoever wants to update the implementation needs to magically know to update the input validation as well. It's an unnecessary burden.

---

Fourth, this change adds unnecessary maintenance overhead for little practical value (see my first point above). As a maintainer, which option would you choose?

1. Old implementation, 99% correct, one line of code, covered by unit tests
2. New implementation, 99% correct, 50 new lines of code, 100 new lines of tests

---

I hope I've conveyed my position clearly.

Reply:
Thank you @mpenkov  for such an interesting discussion! First of all, you still did not convince me and my opinion is not changed, **but** I will revert the code - it is not worth writing a paper here, and you are the maintainer :)

Second, you are commenting on the old version. Here is the current one:
<details>
<pre>
def _check_stream(fileobj, mode):
    """
    Detect whether the specified object is a file object with required capabilities
    implied by `mode`. Raise a TypeError if the file object capabilities are insufficient.

    Return True if `fileobj` is a file object; otherwise, False.
    """
    has_read = hasattr(fileobj, 'read')
    has_write = hasattr(fileobj, 'write')
    if not has_read and not has_write:
        return False
    # otherwise, we consider that we have a file object
    # do we have the required methods?
    # we always do for open()-like file objects, however, it may not be the case with custom types
    if mode.endswith('+') and (not has_read or not has_write):
        raise TypeError(
            'File object-like %r does not have the "read" or "write" methods required by mode %s' % (fileobj, mode))
    if mode[0] == 'r' and not has_read:
        raise TypeError('File object-like %r does not have the "read" method required by mode %s' % (fileobj, mode))
    if mode[0] in ('w', 'a') and not has_write:
        raise TypeError('File object-like %r does not have the "write" method required by mode %s' % (fileobj, mode))
    # ok, methods are here, but we need to further check the mode of open()-like file objects
    try:
        open_mode = getattr(fileobj, 'mode')
    except AttributeError:
        # this is not a file object derived from open(), let's hope for the best
        return True
    if (mode.endswith('+') and not open_mode.endswith('+')) or \
            (not open_mode.endswith('+') and open_mode[0] != mode[0]):
        raise TypeError("%r was opened in an incompatible mode: need %s have %s" % (fileobj, mode, open_mode))
    return True
</pre>
</details>

Below is my view on your points.

### Practical valueThere is a big difference between getting a raw exception and a targetted exception. Compare:

```
io.UnsupportedOperation: not writable
```

and

```python
<_io.TextIOWrapper name='setup.py' mode='r' encoding='UTF-8'> was opened in an incompatible mode: need wb have r
```

We see at once that:

* We opened "setup.py"
* We've got a conflict between the modes

I do understand that you may be used to a different way of developing software, and `smart_open` may have well-educated, capable for debugging users. This is my (somewhat long) experience.

> More specifically, the new code is spending an unnecessary amount of effort preparing for edge cases that do not occur in practice.

Again: please, comprehend me, I already wrote it: the standard streams **is not** the target goal. Custom classes are.

### Lack of correctness

100% of the cases are not checked, of course. So are the assertions in any unit tests, for example.  One targets covering the most important points first, or those which are easily to cover. Rome was not build in a day, and iteration by iteration, this code may be improved in the future.

I also refer to programming by contract. The old way of specifying the contract was implicit: it breaks where it breaks. The new way is definitely explicit. Now let's read Zen of Python, no. 2.
You can still argue whether it is Pythonic or not though - it is quite subjective. And after all, "practicality beats purity".

### Introduction of unnecessary coupling

I don't know what to say about your understanding of coupling. I will ask! What's your attitude to PEP 484 (type hints) and mypy in particular?

### Disproportionate maintenance overhead

This is hard to oppose because an absense of code is the best code, of course. However, I don't agree with the "correctness" thesis: the lack of features is also correct by definition, and I am adding a new feature.

So happy that I am doing open source! Thanks once again for a great discussion. Maybe I will convert it to a blog post - I know some folks who may be interseted.

Reply:
> Again: please, comprehend me, I already wrote it: the standard streams **is not** the target goal. Custom classes are.

_Properly implemented_ custom I/O classes will behave _identically_ to standard streams, because they will inherit from the [io.IOBase](https://docs.python.org/2/library/io.html#i-o-base-classes) abstract class (or something similar from the io module). That's literally what the entire io module was made for. Now,

> There should be one-- and preferably only one --obvious way to do it.

and the io stdlib is that way, in this case.

That narrows your target goal even further: now you're targeting custom classes that don't inherit from stdlib IO classes (for whatever reason), and don't even bother to implement the standard interface. It's nice that you're being considerate and trying to handle such implementations, but you're preparing for a problem that doesn't exist yet, and your effort essentially duplicates the IOBase abstract class.

> I also refer to programming by contract. The old way of specifying the contract was implicit: it breaks where it breaks. The new way is definitely explicit. 
> Now let's read Zen of Python, no. 2.

Don't stop reading at point no. 2, no. 3 and 4 are no less important ;)

> What's your attitude to PEP 484 (type hints) and mypy in particular?

I'm indifferent to type hints and mypy. I don't suffer from the problems that they solve, so I'm not in a rush to adopt them.

> So happy that I am doing open source! Thanks once again for a great discussion. Maybe I will convert it to a blog post - I know some folks who may be interseted.

Glad you're enjoying the experience. Again, thank you for your work and the discussion. Be sure to let me know how the blog post goes.